### DISCUSSION TITLE
Set password expiration policy

### DISCUSSION THREAD

**Discussion by tpemeja** [authelia/authelia#8560] [Ideas]
See [this comment](https://github.com/authelia/authelia/discussions/8560?converting=1#discussioncomment-11730835) for what you're voting on. Voting for this idea means clicking the up arrow below the start of this discussion, not any below comment.

### Description

Implement a password expiration policy within Authelia to enhance account security by requiring users to change their passwords periodically.

#### 1. System Administrator Configures Policy:
The system administrator configures Authelia with a password expiration policy, specifying parameters such as expiration interval (e.g., 90 days), grace period for password changes, and notification settings.

#### 2. Authelia Monitors Password Ages:
Authelia continuously tracks the age of each user’s password based on the last update timestamp.

#### 3. Notification to Users:
Authelia sends periodic reminders (e.g., 14 days, 7 days, and 1 day before expiration) to users whose passwords are nearing expiration.

#### 4. User Changes Password:
The user receives a notification and logs into the system to update their password before the expiration date.

#### 5. Password Successfully Updated:
Authelia validates the new password according to security guidelines (e.g., complexity requirements) and updates the user’s password record. The expiration counter resets.

#### 6. Access Maintained:
The user continues to access services without interruption after updating the password.

### Details

The administrator sets the following parameters:

- **Expiration Interval:** Defines how long a password is valid before expiring (e.g., 90 days).
- **Grace Period:** Specifies the time allowed for a user to change their password after it has expired (e.g., 7 days).
- **Notification Settings:** Configures the schedule and content of notifications sent to users (e.g., email reminders 14 days, 7 days, and 1 day before expiration).
- **Security Requirements:** Sets the password complexity rules (e.g., minimum length, inclusion of special characters, numbers, and mixed case).
- **Action on Expiration:** Defines what happens when a password expires (e.g., restrict access, redirect to a password reset page).

### Configuration Example:
```yaml
password_policy:
  expiration_days: 90
  grace_period_days: 7
  notification_schedule:
    - 14 # days before expiration
    - 7
    - 1
  action_on_expiration:
    restrict_access: true
    redirect_to_reset_page: true
    display_warning: false
```

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] I have checked for related issues and checked the documentation

**Comment by Crowley723** [comment]
While I like the idea of allowing admins to choose for themselves, I don't think it's the best idea to allow forcing users to change their password arbitrarily.

If a user is required to change their password after a set period of time, they are incentivised to choose a more memorable and likely weaker password then they may choose otherwise.

I've talked to @james-d-elliott about it and he suggested password expiry based off the entropy of a given password.

The zxcvbn strength estimator would be a good way to implement this. The score given when the password is changed/created could set expiry for lower score passwords while passwords at higher scores wouldn't be subject to expiry.

**Comment by james-d-elliott** [comment]
Thanks for the idea.

Effectively the general consensus is that password expiration is a security anti-pattern. While a consensus is not a reasonable cause for conclusion it is still an important factor.

Password expiration and complexity requirements (though, these effects occur regardless of complexity requirements) have the following negative security outcome when it comes to users:

1. Users generally look for methods to circumvent password expiration. Most systems that implement this end up recording password history for this reason. Users will change it to previously used passwords.
2. Users are more likely to write passwords down.
3. Users are more likely to utilize easy to break passwords if they are able, or if they are not, they're likely to use iterations of previous passwords.
4. Users who would otherwise choose a **_good  password_** can lean towards choosing a **_bad password_** because of frustration with the process, because they are effectively being punished for using a **_good  password_**.
 
As such if we do implement something like this we will impose the arbitrary way it's determined if a user is subject to password expiration. An admin would have the choice to turn it on and pick some kind of frequency, but there would be an arbitrary process by which we determine a user should not be subject to password expiration in the case of a **_good  password_**. I think weighting the frequency against a password entropy measurement (to infinity) would be the best outcome. i.e. if the admin picks x frequency, the entropy beyond a certain point would start to multiply that frequency, and eventually become infinity.

**Comment by tecosaur** [comment]
I think it's worth noting that https://pages.nist.gov/800-63-4/sp800-63b.html#passwordver now contains:

> 6. Verifiers and CSPs SHALL NOT require users to change passwords periodically. However, verifiers SHALL force a change if there is evidence of compromise of the authenticator

**Comment by james-d-elliott** [comment]
Thanks for including this. This is one of the many sources which have come to similar conclusions about password expiration which form the basis of my current view of this.

**Comment by tecosaur** [comment]
I'm always happy to push for less annoying + unhelpful security practices :smile:.

On an unrelated note that I just want to make _somewhere_, I just started using Authelia a few days ago and it's been brilliant so far :heart_eyes:. The passkey work you're lining up for 4.39 is great to see too. Thanks for all the work you've put into this, it really shows.

**Comment by james-d-elliott** [comment]
For the reasons mentioned above we're going to convert this into an idea which people can vote on. In general this means we're soft rejecting it until we know enough people want it (so if you want it, please vote so we know you want it). 

The idea specifically being voted on is as follows:

1. The policy itself will be off by default as per the current NIST compliance guidelines and general recommendations.
2. Authelia will store a metric of the entropy level a users password has at the time they change it.
3. Authelia will store a timestamp of the last password change. Any user who has not changed their password will not have this timestamp and will be ignored by the policy but we will give a means to set a list of users to the current timestamp. 
3. Authelia will utilize the entropy measurement to do the following with no option to configure this behavior:
    1. Weight the configured frequency against the entropy measurement. This will mean stronger passwords will naturally have longer expiration.
    2. Decide if the measurement is sufficient enough to disable the policy entirely for the user.
4. Probably require admins to accept a disclosure about the problematic nature of enabling it (probably means only way to enable it is via CLI).

**Comment by agent-ck** [comment]
Not to bore with an old topic, but there is another use case for forcing users to change their passwords. 

What about new users who have been configured with an initial password but are not supposed to keep it? For those users, a forced password change *is* urgently necessary!

Or is there an alternative configuration, I haven't found yet?

**Reply by Crowley723** [reply]
It's likely that we will address that use case either in the user management feature or after it.

I would consider this a different feature request from the above issue. The above issue discusses **automatically** requiring password changes on a schedule, your use case would likely be part of the onboarding/registration process (potentially once).

**Reply by james-d-elliott** [reply]
As mentioned this is a different idea. In general I'd suggest opening a separate idea discussion so users can vote on it and we can ascertain the level of importance based on user votes. We would have to invest considerable time into a feature like this as every LDAP backend has its own way to indicate the user must change a password on next login.

Realistically also, just setting the password to a completely random value achieves the same outcome. A feature like this will require the user to use the existing reset password functionality which requires SMTP and a valid email address. The only realistic difference is user and admin experience, not effective functionality.

**Reply by agent-ck** [reply]
It was not my intention to start a new FR. I just wanted to show that there can be some valid reasons for enforcing password changes, regardless of the process in which this step is implemented.  

Sure, it's easier with an LDAP backend, but there's also user management in the yaml file. Of course, I can make manual entries there, but I'm opposed to doing so for those reasons:  

1. After changes, the service must be restarted. A reload is not possible.
2. The possibility of errors increases significantly, as is always the case with manual interventions in configurations. 
3. When the yaml is incorrect, the service won't start and becomes unavailable, until the yaml is fixed.

Is that really more secure than implementing a CLI command?

As I already wrote, I just wanted to provide a valid use case. So you don't have to reply to me, and I don't really expect this to result in any changes. My aim was more to highlight these points so that they can perhaps be remembered in future roadmaps. Nothing more, but also nothing less!

I worked around this problem by using multiple instances of Authelia. One for development, one for testing automation processes, and finally the production instance, which is only changed by automation, when the test instance can go online after the automated changes.

**Reply by james-d-elliott** [reply]
> After changes, the service must be restarted. A reload is not possible.

I don't think this is exactly accurate. But maybe if you elaborate on what you mean.

> I just wanted to show that there can be some valid reasons for enforcing password changes, regardless of the process in which this step is implemented.

Thanks, appreciate your follow up and the dialogue. Several factors you have brought up are realistically a future consideration. I think in general there are reasons to implement measures to enforce a password change at next login. That is not to say we absolutely should, but it's something we will endeavor to consider in the future (effectively why I suggested another idea discussion).

There needs to be a fairly reasonable consideration to the long term maintainability of something like this. Several LDAP backends have their own way of communicating that a user must change their password at next login for example, and realistically we should endeavor to make that experience seamless for users.

We have to be incredibly careful about muddying the storage of information about users. Credentials make a lot of sense to store in the database, or UI preferences. But storing additional information that realistically should be stored in the authentication backend is a bell that's incredibly difficult to unring.

**Comment by slawekgl79** [comment]
Hi there, i am evaluating Authelia for our company, but this topic is from our perspective a blocker. 

For example, PCI DSS 4.0 still explicitly requires password rotation (every 90 days in many setups). In practice, a lot of companies stick to that for audit reasons.

So if there’s no way to enforce password expiration directly, it can make Authelia hard to use in PCI‑certified setups unless you handle it elsewhere (LDAP/IdP, etc.).

I know modern guidance (like NIST) moves away from forced rotation, but from a compliance/audit perspective it’s still a hard requirement in some cases.

**Reply by james-d-elliott** [reply]
Are you not using MFA?

**Reply by slawekgl79** [reply]
> Are you not using MFA?

At the PCI DSS–certified company where I work, we use MFA, as required by PCI DSS, along with password rotation policies.
On my test setup, I’ve been evaluating Authelia. I haven’t implemented MFA there yet. I did configure a few simple scenarios with reverse proxies, which worked well, but when I tried to introduce password policies, I got stuck and paused my setup at that point.

**Reply by nightah** [reply]
The question was directly related to the PCI DSS requirement: if MFA is used for user accounts, the 90-day password rotation isn't necessary. It's really only system/application accounts that need to be considered.

You should discuss the specifics of your environment and controls with your QSA, but the spec is quite clear in what is expected.

<img width="1684" height="1104" alt="image" src="https://github.com/user-attachments/assets/74214c29-d6c3-4b0c-bb51-69ae6d8a11d2" />

**Reply by james-d-elliott** [reply]
The requirement in regards to password rotations in PCI DSS 4.0 vs 3.2.1 was that they clarified what meets compliance. In particular this note in the changelog:

> Clarified that this requirement applies if passwords/passphrases are used as the only authentication factor for user access (i.e., in any single-factor authentication implementation)

This is detailed in requirement 8.3.9.

**Reply by slawekgl79** [reply]
Thanks a lot for the detailed explanation — really appreciate it!

I’ll double-check this with our QSA to be 100% sure, but it looks like I missed a fairly major change that came with the PCI DSS 4.0 update around password rotation vs. MFA.

Also, I just wanted to say that I really like Authelia — especially the YAML-based configuration. It makes it super straightforward to deploy and manage consistently using tools like Puppet or Ansible, which fits very well into our workflow.
Thanks again for the help 👍

**Reply by james-d-elliott** [reply]
No problem glad we could help. Please let us know if you don't have success with the QSA, or in general.

We put a lot of effort into security conscious defaults from the first implementation. We can provide specific guidance on which options may need to be tweaked for the best / inarguable compliance with this specific or other specific requirements.
